File: guides/openai-with-sandbox/assistant.py
# Import libraries
import os
import e2b
import openai
from openai import OpenAI

# Set up API key
from dotenv import load_dotenv
import json

load_dotenv()
openai.api_key = os.environ["OPENAI_API_KEY"]


# Optional - import types, allowing for explicit type hints
from typing import Any, Dict


from typing import List
from openai.types.beta.assistant_create_params import Tool


def create_assistant():
    client = openai.Client()
    
    functions: List[Tool] = [
        {
            "type": "function",
            "function": {
                "name": "save_code_to_file",
                "description": "Save code to file",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "code": {
                            "type": "string",
                            "description": "The code to save",
                        },
                        "filename": {
                            "type": "string",
                            "description": "The filename including the path and extension",
                        },
                    },
                },
            },
        },

    ]

    # current_directory = os.getcwd()

    ai_developer = client.beta.assistants.create(
        instructions="""You are an AI developer.

    When given a coding task, write and save code to files and install any packages if needed.
    Start by listing all files inside the repo. You work inside the '/home/user/repo' directory.

    Please print any code that you have written there also to the terminal.

    Thank you, you're the best!
    """,
        name="AI Developer",
        tools=functions,
        model="gpt-4-1106-preview",
    )

    print("AI Developer Assistant created, please copy its ID below to your .env file. You can find all your created assistants at https://platform.openai.com/assistants.")
    print(ai_developer.id)


if __name__ == "__main__":
    create_assistant()

File: examples/openai-assistant-js/assistant.ts
import { AssistantCreateParams } from 'openai/src/resources/beta/assistants/assistants';
import OpenAI from 'openai';
import 'dotenv/config';

const openai = new OpenAI();

export const functions: Array<
  | AssistantCreateParams.AssistantToolsCode
  | AssistantCreateParams.AssistantToolsRetrieval
  | AssistantCreateParams.AssistantToolsFunction
> = [
  {
    type: 'function',
    function: {
      name: 'saveContentToFile', // Updated action name
      description: 'Save content (code or text) to file',
      parameters: {
        type: 'object',
        properties: {
          content: {
            type: 'string',
            description: 'The content to save',
          },
          filename: {
            type: 'string',
            description: 'The filename including the path and extension',
          },
        },
      },
    },
  },
  {
    type: 'function',
    function: {
      name: 'writeToFile', // Updated action name
      description: 'Write text or other content to a file',
      parameters: {
        type: 'object',
        properties: {
          content: {
            type: 'string',
            description: 'The content to save',
          },
          filename: {
            type: 'string',
            description: 'The filename including the path and extension',
          },
        },
      },
    },
  },
  {
    type: 'function',
    function: {
      name: 'listFiles',
      description: 'List files in a directory',
      parameters: {
        type: 'object',
        properties: {
          path: {
            type: 'string',
            description: 'The path to the directory',
          },
        },
      },
    },
  },
  {
    type: 'function',
    function: {
      name: 'readFile',
      description: 'Read a file',
      parameters: {
        type: 'object',
        properties: {
          path: {
            type: 'string',
            description: 'The path to the file',
          },
        },
      },
    },
  },
  {
    type: 'function',
    function: {
      name: 'commitAndPush',
      description: 'Commit and push changes to the repository',
      parameters: {
        type: 'object',
        properties: {
          commitMessage: { // Updated parameter name
            type: 'string',
            description: 'The commit message',
          },
        },
      },
    },
  },
];

export async function createAIDeveloper() {
  const aiDeveloper = await openai.beta.assistants.create({
    instructions: `You are an AI developer.
    The provided codebase is in the /home/user/repo.
    When given a coding task, you will work on it until it is completed. You will summarize your steps.
    If you encounter some problem, just communicate it please. 
    You can save code to file (or create a new file), list files in a given directory, read files, and commit and push changes.
    Please every time you are asked to do a task, do the task the best you can, and then commit and push it without asking.
    You are professional, don't argue, and just complete the task.`,
    name: 'AI Developer',
    tools: [...functions],
    model: 'gpt-4-1106-preview',
  });

  console.log('AI Developer Assistant created, copy its id to .env file:');
  console.log(aiDeveloper.id);
}

if (import.meta.main) {
  createAIDeveloper();
}
